flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mongodb-cdc/src/main/java/org/apache/flink/cdc/connectors/mongodb/source/reader/fetch/MongoDBStreamFetchTask.java

```diff
@@ -108,7 +110,32 @@ public void execute(Context context) throws Exception {
         this.taskRunning = true;
         try {
             while (taskRunning) {
-                Optional<BsonDocument> next = Optional.ofNullable(changeStreamCursor.tryNext());
+                Optional<BsonDocument> next;
+                try {
+                    next = Optional.ofNullable(changeStreamCursor.tryNext());
+                } catch (MongoQueryException e) {
+                    if (e.getErrorCode() == CHANGE_STREAM_FATAL_ERROR) {
+                        ChangeStreamOffset offset =
+                                new ChangeStreamOffset(streamSplit.getStartingOffset().getOffset());
```

Comment:
Thanks @yuxiqian for this fix.
Why don't we catch this exception when we are in `openChangeStreamCursor`? 
Is there any consideration?

Reply:
I checked the following stack trace in #1879, and this exception occurs when opening the cursor.

```
at com.mongodb.internal.operation.OperationHelper.withReadConnectionSource(OperationHelper.java:583)
at com.mongodb.internal.operation.CommandOperationHelper.executeCommand(CommandOperationHelper.java:189)
at com.mongodb.internal.operation.AggregateOperationImpl.execute(AggregateOperationImpl.java:195)
at com.mongodb.internal.operation.ChangeStreamOperation$1.call(ChangeStreamOperation.java:323)
at com.mongodb.internal.operation.ChangeStreamOperation$1.call(ChangeStreamOperation.java:319)
at com.mongodb.internal.operation.OperationHelper.withReadConnectionSource(OperationHelper.java:583)
at com.mongodb.internal.operation.ChangeStreamOperation.execute(ChangeStreamOperation.java:319)
at com.mongodb.internal.operation.ChangeStreamOperation.execute(ChangeStreamOperation.java:58)
at com.mongodb.client.internal.MongoClientDelegate$DelegateOperationExecutor.execute(MongoClientDelegate.java:184)
at com.mongodb.client.internal.ChangeStreamIterableImpl.execute(ChangeStreamIterableImpl.java:204)
at com.mongodb.client.internal.ChangeStreamIterableImpl.access$000(ChangeStreamIterableImpl.java:53)
at com.mongodb.client.internal.ChangeStreamIterableImpl$1.cursor(ChangeStreamIterableImpl.java:129)
at com.mongodb.client.internal.ChangeStreamIterableImpl$1.cursor(ChangeStreamIterableImpl.java:121)
at com.ververica.cdc.connectors.mongodb.source.reader.fetch.MongoDBStreamFetchTask.openChangeStreamCursor(MongoDBStreamFetchTask.java:237)
... 8 more
when use mongodb cdc (sql) transfer the mongodb table to flink table store table，if enable the scan.incremental.snapshot, the job will appear the error: " the resume token was not found" immediately.
```

Reply:
I think we can refer to the logic in mongo-kafka to reopen the changestream.

https://github.com/mongodb/mongo-kafka/blob/2ffd620441b68bec19227dbfa641375d37148e39/src/main/java/com/mongodb/kafka/connect/source/StartedMongoSourceTask.java#L102

Reply:
Yes, it would be much cleaner to move that checking logic to `openChangeStreamCursor` and implement the fallback logic along with existing routes.

CMIIW but seems `MongoCommandException` will only be thrown [when getNext or tryGetNext is called](https://github.com/mongodb/mongo/blob/a71feef352a1f08e3f9dbc2b840691e75775a370/src/mongo/db/pipeline/document_source_change_stream_ensure_resume_token_present.cpp#L133C31-L133C83), and we can't intentionally trigger an `MongoCommandException` by calling other methods like `hasNext` earlier.

@Jiabao-Sun Any suggestions about this?

Reply:
Hi @yuxiqian, the `MongoCommandException` seems can be thrown by openChangeStreamCursor.

```
at com.mongodb.client.internal.ChangeStreamIterableImpl$1.cursor(ChangeStreamIterableImpl.java:121)
at com.ververica.cdc.connectors.mongodb.source.reader.fetch.MongoDBStreamFetchTask.openChangeStreamCursor(MongoDBStreamFetchTask.java:237)
```

![image](https://github.com/apache/flink-cdc/assets/27403841/829b5668-1cbf-4bfe-a324-a05c53e8d2eb)

Reply:
@Jiabao-Sun You're right, I've added resume token expiration checks in creating and iterating process.